### DISCUSSION TITLE
Information about pypdfium2 functionalities for working with PDF text

### DISCUSSION THREAD

**Discussion by EmanueleGusso** [pypdfium2-team/pypdfium2#88] [Q&A] [answered]
Dear @adam-huganir and @mara004 
First of all I'd like to thank you for your amazing work.
I'm looking for a library which is able to identify the bounding boxes of characters and text lines in a pdf in order to obtain:
1. the coordinates of each character;
2. the coordinates of each text line ordered in reading order;
3. the textual content of the text lines.

This functionalities are provided, for example, by https://github.com/pdfminer/pdfminer.six but I am looking for an alternative and I read about foxit and then I read about you.
Reading the foxit documentation (http://cdn01.foxitsoftware.com/pub/foxit/manual/enu/FoxitPDF_SDK20_Guide.pdf?fbclid=IwAR0jfV9RyMx2pgBzboe8YRnPwoD98Fdp1XNYQBn-BGOXev3K4V-kJEDV7iQ) it seems like the three features I mentioned are provided but it's not clear whether foxit is free or not. 
So I'm writing you to ask if pypdfium2 is able to fulfill those 3 requirements.
I really hope you can answer me and I really thank you in advance for your availability.
Emanuele

**Comment by mara004** [comment]
Dear @EmanueleGusso,

Thanks for your interest in pypdfium2. I haven't used the text functions of PDFium yet, but since Chromium is able to mark and extract text, I would presume that it's possible to get all required information using the low-level ctypes interface of pypdfium2. I think you may be looking for the [`FPDFText_...()`](https://pdfium.googlesource.com/pdfium/+/refs/heads/main/public/fpdf_text.h) functions?
1. Probably [`FPDFText_GetCharBox()`](https://developers.foxit.com/resources/pdf-sdk/c_api_reference_pdfium/group___f_p_d_f_i_u_m.html#ga3571462e0c83e2472f378c06582bdd69).
2. [`FPDFText_GetRect()`](https://developers.foxit.com/resources/pdf-sdk/c_api_reference_pdfium/group___f_p_d_f_i_u_m.html#gad2ff37d1eb1704a0654cd26070d57bae). (However, I belive this is based on heuristics, since PDFs only store the position of each text glpyh, but there's no information on lines or text groups.)
3. [`FPDFText_GetBoundedText()`](https://developers.foxit.com/resources/pdf-sdk/c_api_reference_pdfium/group___f_p_d_f_i_u_m.html#ga211cb3a6035142ab390f75a018decfc9) may be used to obtain text of a certain area on a page.

I recommend that you take a look at [`pdf_text_page.py`](https://github.com/innodatalabs/pdfbrain/blob/master/pdfbrain/pdf_text_page.py) of [pdfbrain](https://github.com/innodatalabs/pdfbrain), which implements a support model for working with the PDFium text API.

Anything of [this section](https://developers.foxit.com/resources/pdf-sdk/c_api_reference_pdfium/group___f_p_d_f_i_u_m.html) in the Foxit documentation is freely available in the open-source PDFium. That said, I think pdfminer.six has an excellent reputation for tasks related to PDF texts. Why are you looking for an alternative, if I may ask?

**Reply by EmanueleGusso** [reply]
Dear @mara004 
First of all thank you very much for your answer. You clarified a few things and I'll have a look at pdfbrain before trying to use directly pdfium because I'd rather stick with the Python language.
You're right, pdfminer has an excellent reputation but I am looking for the most reliable library and I found some issues with pdfminer (like this one https://github.com/pdfminer/pdfminer.six/issues/281 ).
So, I'm afraid that there are some pdfs which, for unknown reasons, aren't properly handled by pdfminer.
When I read about PDFium I read that it is now maintained by Google so I immediately thought that PDFium should be the most reliable thing out there. What do you think about this thought? 
For sure you are more experienced than me on this subject so I am open to all your suggestions.
Again, thank you very much for your work and your availability.

**Reply by mara004** [reply]
Dear @EmanueleGusso,

> You clarified a few things and I'll have a look at pdfbrain before trying to use directly pdfium because I'd rather stick with the Python language.

Well, you could still use pypdfium2 & Python, but the referenced code of pdfbrain should give you a good idea on how to interface with the mentioned PDFium API functions using [`ctypes`](https://docs.python.org/3.8/library/ctypes.html). Let me know if you have any problems with this, and I'll try to help (though my knowledge of ctypes is quite limited yet).

> When I read about PDFium I read that it is now maintained by Google so I immediately thought that PDFium should be the most reliable thing out there. What do you think about this thought?

I'd recommend that you try opening some of your sample documents (on which pdfminer.six does not work properly) in Chromium to see if PDFium detects text positions correctly.

**Comment by mara004** [comment, marked as answer]
I have now added a support model for text extraction with commits https://github.com/pypdfium2-team/pypdfium2/commit/bbc2438f0189ae3118fa3aacdaf6120f2827a72b and https://github.com/pypdfium2-team/pypdfium2/commit/58396a2b5566e5538886a838695607bf858dfdfd.
